plugins/sinks/kafka/sink.go

```diff
@@ -146,14 +152,46 @@ func (s *Sink) extractKeyFromPayload(fieldName string, value interface{}) (strin
 func (s *Sink) getTopLevelKeyFromPath(keyPath string) (string, error) {
 	keyPaths := strings.Split(keyPath, ".")
 	if len(keyPaths) < 2 {
-		return "", errors.New("invalid path, require at least one field name e.g.: .URN")
+		return "", errors.New("invalid path, require at least one field name e.g.: .Urn")
 	}
 	if len(keyPaths) > 2 {
 		return "", errors.New("invalid path, doesn't support nested field names yet")
 	}
 	return keyPaths[1], nil
 }
 
+func castModelToPointer(value interface{}) interface{} {
```

Comment:
can you help me understand what are we doing here and if there is any way we can avoid this?

Reply:
@kushsharma  So basically when we are casting the payload (type `interface{}`) to proto message like below
e.g.
```
payload.(proto.Message)
```
it will always fail because the model only implements `proto.Message` when it is a pointer and currently we are working with struct instead of a pointer.

```
# meta.Table
func (*Table) ProtoMessage() {}
```


It is not a graceful solution but other alternatives are:
 1. change all of our logic to use our model as a pointer instead of a struct
 2. build an intermediate model for each of our models to handle proto serializing (can even use embedded struct)

I personally do not like the first approach as im not sure if this specific case is a good reason to use a pointer.
The second approach looks better in my opinion as it does makes sense for meteor to have its own model, but needs more discussion on how to do it.

And as for the current approach, Im just assuming we have a model and we need to map all the fields to its respective proto class before serializing it into protobuf, only in this case it is much simpler as we just need to turn it into pointer.

Reply:
@StewartJingga pulled the changes locally and tested it with goreleaser as well, seems fine to me.

Reply:
Hi @kushsharma, any thoughts on this?

Reply:
Sorry for the late reply just saw the comment.
Why are we sending the struct in the first place instead of reference? I think we should be sending the pointer anyway right? So the example I wrote earlier with `date->kafka`, I used a struct reference to demonstrate this. That is your first solution. `1. change all of our logic to use our model as a pointer instead of a struct` 

Can you elaborate on this a little? `I personally do not like the first approach as I'm not sure if this specific case is a good reason to use a pointer.` Also, I wonder how is this simpler? All of our sinks need to type-cast this to check its type, you won't be able to pass a struct that implements an interface that the sinks can use later like `ToJSON` or `ToProto` or something as well.

Reply:
Also, you can always get the element from reference but I am not sure it goes the other way around in reflect library as well.

Reply:
@kushsharma 
> Can you elaborate on this a little? I personally do not like the first approach as I'm not sure if this specific case is a good reason to use a pointer.

So basically I always go for struct as the first option when passing data for immutability purposes.
And if we are using pointer, there is some kind of pointless-ness in sending data from channels to channels as they are all actually pointing to the same data.

> Also, I wonder how is this simpler?

Imagine, if actually Meteor's internal models are not a proto model but instead just plain golang structs.
And then you are trying to build a kafka sink to send them as proto bytes. Then we would need to build a mapper to translate each of the models to their respective proto before marshaling them into bytes.

When I said simpler, I was comparing if we have to map the fields from plain struct to proto struct, to our case now which is just mapping it to pointer.

> All of our sinks need to type-cast this to check its type

Actually just this sink, because of proto marshalling purpose.

Reply:
> So basically I always go for struct as the first option when passing data for immutability purposes.
And if we are using pointer, there is some kind of pointless-ness in sending data from channels to channels as they are all actually pointing to the same data.

Yeah, I agree we should always use struct unless necessary but the current use case clearly requires us to do so. Also because we are dealing with a large amount of data it's actually better to pass it as a pointer because a single struct message could be as big as multiple MBs, I think in the worst case, and passing it as a struct will require the compiler to create object copies every time. I agree immutable objects are nice but here the tradeoff is too big.

> Imagine, if actually Meteor's internal models are not a proto model but instead just plain golang structs.
And then you are trying to build a kafka sink to send them as proto bytes. Then we would need to build a mapper to translate each of the models to their respective proto before marshaling them into bytes.

Exactly! that's what we want to avoid right? Maintaining a mapper would be troublesome, if there is a way to avoid it then we should always do it. If there was no other way to do it then I would agree with you but here we clearly have a solution to avoid it.

> Actually just this sink, because of proto marshalling purpose.

You are missing the point, don't treat this as proto marshalling only, here we are asserting the underlying interface that implements proto marshalling, it could have been JSON as well or something else `ToXYZ()`. The point is each sink that we create will have to maintain this mapping to work properly and whenever someone starts producing a new type of extracted struct, the sink needs to again update itself, not just one, all sinks needs to update. I see no real advantage in this approach compared to passing a pointer.

Reply:
We can take opinions from others as well I guess? I could be wrong 😄 
@ravisuhag @harikrishnakanchi @turtleDev

Reply:
> Also because we are dealing with a large amount of data it's actually better to pass it as a pointer because a single struct message could be as big as multiple MBs

This is a pretty good case for a pointer 👍 

> The point is each sink that we create will have to maintain this mapping to work properly and whenever someone starts producing a new type of extracted struct, the sink needs to again update itself, not just one, all sinks needs to update.

I'm still unable to see this 🤔, can you give me an example? This is just because the proto message interface implementation is on the pointer which i'm sure pretty specific case and won't happen elsewhere.

The way I see it is just like how I see other applications when they want to publish protobuf. They will, inevitably, have to build a mapper to map their models to respective protos and this time we are doing just that.

Reply:
> You are missing the point, don't treat this as proto marshalling only, here we are asserting the underlying interface that implements proto marshalling, it could have been JSON as well or something else ToXYZ().

unless the sink requires proto marshalling, there won't be a need to check the underlying interface.

And in case multiple sinks requires this mapping, we could create an adapter module to map our model to proto (or anything) and the heavylifting will be done and centralised there.

Reply:
I'm going through the discussion, and just trying to sum up.

1. the current approach solves the use-case of proto serialisation
2. it doesn't scale well because we have to update this everytime a new model is added
3. since we're passing data via copying, it's going to have added over head

assuming all the problems above;
2. can be solved by using `reflect.Pointer` API to dynamically return a pointer, rather than having a type switch. This way we don't need to change it. We can also move this to the serialiser (in case it isn't a part of it already, haven't read the rest of the code) to tidy it up.
3. Golang's memory model uses copy-on-write semantics, so we don't necessarily need to worry about moving large data by value, as long as we don't mutate it a lot. As always, we can benchmark and find out.

Reply:
You can essentially rewrite this function as:

```golang
func CastModelToPointer(v interface{}) interface{} {
    return reflect.ValueOf(v).Addr().Interface()
}
```

Reply:
About the discussion about finding a common interface any type an implement, +1. But with the nature of this application, finding the correct contract might be a challange.__

Reply:
@turtleDev tried your approach but still got issue as apparently the value is not addressable hence panic-ing on `Addr()`.

But I did find another approach to solve this, it includes another copying but should simplify the mapping.

```
func castModelToPointer(value interface{}) interface{} {
	vp := reflect.New(reflect.TypeOf(value))
	vp.Elem().Set(reflect.ValueOf(value))

	return vp.Interface()
}
```

wdyt?